sympy/tensor/indexed.py

```diff
@@ -359,7 +259,10 @@ def args(self):
             return self._args
 
     def _hashable_content(self):
-        return Expr._hashable_content(self) + (self._shape,)
+        if self._shape:
+            return super(IndexedBase, self)._hashable_content() + (self._shape,)
+        else:
+            return super(IndexedBase, self)._hashable_content()
```

Comment:
Hmm this was actually a bug fix, not an optimization... If the `IndexedBase` doesn't have a shape then `_shape` is `None` (from other code not written by me!). So the original code adds `None` to `_hashable_content` which then created errors down the line in rare situations IIRC. Maybe I'm misunderstanding what you're asking for, though?